### PR TITLE
Continuous Space Pathfinding

### PR DESCRIPTION
**Pull request by AayushSabharwal**
None of the new methods are tested, and the docstrings are just copied over from `astar_grid.jl`. ~I'll work on getting `move_along_route!` working for periodic continuous spaces before addressing that~ (done). Some points to address:
- `set_target!` and `set_best_target!` for continuous space require the model to be passed in, for continuous space dimensions
- I chose to keep `find_path` as it is and add a new `find_continuous_path` method. The alternative was having `find_path` return `parent` and each `set_target!` method extract the path from that, possibly through another method such as `unravel_path`.

TODO:
- [x] Revert `ci.yml`
- [x] Tests
- [x] Docs
- [x] Example

### CODE REVIEW

**Comment by AayushSabharwal** [comment]
Does CI not run if the PR isn't to master?

**Comment by Libbum** [comment]
[Looks like it](https://github.com/JuliaDynamics/Agents.jl/blob/4362c80b35c4628ba6be7ee5da239e29c3a4c16e/.github/workflows/ci.yml#L5)...

**Comment by Libbum** [comment]
Feel free to add `pathfinding_rework` into that file for this branch. We'll delete that change before we merge.

**Comment by AayushSabharwal** [comment]
Tests still won't pass right now. It looks like continuous space needs a better metric of length of a path over just counting the number of nodes

**Comment by Datseris** [comment]
> There was an issue with converting to discrete coordinates 

perhaps you can take advantage of the new `get_spatial_index` function I wrote?

**Comment by AayushSabharwal** [comment]
> perhaps you can take advantage of the new `get_spatial_index` function I wrote?

Ah yes, I didn't notice that. I ended up doing the same thing on my own. It's probably a good idea to use `get_spatial_index` and cast the result to `Tuple`, for consistency. I'll merge master into this branch and use it.

**Comment by AayushSabharwal** [comment]
Currently the pathfinding methods all use `::NTuple{D,Float64}` as the type for accepting positions. I noticed that `get_spatial_property` doesn't have such a restriction, and neither do the API methods for `ContinuousSpace`. Should this type restriction be removed from the `astar_continuous.jl` methods as well?

**Comment by AayushSabharwal** [comment]
How should the functions in `astar_continuous.jl` be documented? `move_along_route!` needs its own docstring, but the other two differ only in function signature. Repeating the same or similar docstring seems redundant. So far the options I've come up with are:

- `set_target!` and `set_best_target!` in `astar_continuous.jl` get a one-liner docstring simply saying that this is the function signature for pathfinding in continuous space. `move_along_route!` has its own docstring
- Add a new file `astar_space_api.jl`, which contains definitions such as `set_target!() = nothing` and only serves the purpose of having a docstring. The docstring contains both function signatures, and in the case of `move_along_route!` an additional section for describing the new parameters.

**Comment by AayushSabharwal** [comment]
I also ended up creating a `get_direction` vector for obtaining the vector from one point to another in `ContinuousSpace`, taking into account periodicity. This could potentially be useful outside of pathfinding too. Should it be added as a continuous space exclusive, or some other relevant part of the public API?

**Comment by Libbum** [comment]
> Currently the pathfinding methods all use ::NTuple{D,Float64} as the type for accepting positions. I noticed that get_spatial_property doesn't have such a restriction, and neither do the API methods for ContinuousSpace. Should this type restriction be removed from the astar_continuous.jl methods as well?

What you're using there is fine that's what a position should be in continuous space, so you accept it. There's a fine line between restricting too much, not restricting enough, and being pedantic about all the little helpers. Generally speaking you should provide general methods when you can, only restrict when you know something like `::NTuple{D,Int}` wouldn't work, or it helps dispatch.

I wouldn't worry about doing anything here at the moment.

> How should the functions in astar_continuous.jl be documented?

Just add docstrings where you think they need to be. If the method is exported, Documenter will sort it out. [Take a look at add_agent!](https://juliadynamics.github.io/Agents.jl/stable/api/#Agents.add_agent!) for example: `add_agent!` from two different places are added together.

If you think it's better to have something like [`kill_agent!`](https://juliadynamics.github.io/Agents.jl/stable/api/#Agents.kill_agent!), then in principle you could use your `astar_space_api.jl` trick, but if this is just for docs I don't think it's a good idea. Better to go with your first choice and just make a simple docstring.

> get_direction ...  Should it be added as a continuous space exclusive?

Yes, that'd be a good place.

File: src/submodules/pathfinding/astar_continuous.jl
Comment (Libbum) [inline src/submodules/pathfinding/astar_continuous.jl:97]: If this is to be a ContinuousSpace exclusive, may want to add a `get_direction(from, to, model) = get_direction(from, to, model.space)` wrapper.

**Comment by AayushSabharwal** [comment]
Considering `edistance` exists for `GridSpace` too, I added methods for `get_direction` in `GridSpace` as well. I'm not sure where to put it in docs, though

File: src/spaces/utilities.jl
Comment (Datseris) [inline src/spaces/utilities.jl:60]: Returns -> Return, use imperative.

File: src/spaces/utilities.jl
Comment (Datseris) [inline src/spaces/utilities.jl:76]: This seems like a rather inefficient implementation. Why do you have to make so many vectors? You make a vector with `all_dirs`, you make another vector with `map`. Two alternatives:

1. Loop over possibilities and simply keep track of minimum
2. Just use the `round` function to get the best/proper direction instead of `argmin` of all directions. This doesn't really make sense.

Also, the way this is written, it explicitly assumes 2D-space, yet any dimensional space can be given as an input.

**Comment by Libbum** [comment]
> I added methods for get_direction in GridSpace as well. I'm not sure where to put it in docs, though

Maybe moving agents, together with walk!?

File: src/spaces/utilities.jl
Comment (AayushSabharwal) [reply src/spaces/utilities.jl:76]: Thanks for pointing this out! It looks like I had 2D space in mind when making this and didn't generalize it. I'll rework this function

**Comment by AayushSabharwal** [comment]
Is there any reason `size(space)` isn't defined for `ContinuousSpace`? It would enable avoiding repetition of functions, such as with `get_direction` where nearly the same function is repeated just because the size of the space is obtained differently

**Comment by Datseris** [comment]
just go ahead and define it

**Comment by AayushSabharwal** [comment]
I think docs look good, maybe a little repetitive. There's a couple things I want to test, and then I'll mark this as ready for review

**Comment by AayushSabharwal** [comment]
`nearby_positions` can be used for a random walk in `GridSpace`, but there is currently no equivalent in `ContinuousSpace` if a pathfinder is used. What if `nearby_positions` returned a random position in a random walkable cell of the pathfinder?

**Comment by Datseris** [comment]
>  Is there any reason for it not to be another method of nearby_positions?

That sounds fine to me.

> nearby_positions can be used for a random walk in GridSpace, but there is currently no equivalent in ContinuousSpace 

That is a bit tricky, because the number of nearby positions in `ContunuousSpace` is infinite, no matter how small the radious.

**Comment by AayushSabharwal** [comment]
> That is a bit tricky, because the number of nearby positions in `ContunuousSpace` is infinite, no matter how small the radious.

True. However, computationally it's finite, since we can only store so many significant digits. Similar to how `rand` can only output finitely many values, despite being theoretically infinite. Is there a reason we can't do the same here?

**Comment by Datseris** [comment]
Yes. `rand` outputs a single number, while `nearby_positions` is _an iterator over all positions_. You don't use `rand` to _iterate over all numbers between 0 and 1_. That's why we have the `random_position` function, which is defined for `ContinuousSpace`.

**Comment by AayushSabharwal** [comment]
Ah yes, I got the two things confused. How should random walks be handled in continuous space pathfinding, then? Should I add methods for `walk!`, which call `set_target!` on a random position within a given radius?

**Comment by Datseris** [comment]
I don't see why `walk!` is now interlaced with `set_target!`. The original definition of `walk!` is 
```
walk!(agent, direction::NTuple, model; ifempty = false)
walk!(agent, rand, model)
```
Neither is related to pathfinding. I think I've lost the thread of discussion, can you provide a summary of why `walk!` and pathfinding are related? 

For ContinuousSpace, `walk!` could have a method with a radius:
```
walk!(agent, radius::Real, model{COntinuousSpace})
```
which propagates the agent in a random direction by an amount given by the `radius`?

**Comment by Datseris** [comment]
Actually Im confused by this whole business:

> What if nearby_walkable returned a random position in a random walkable cell of the pathfinder?

Isn't the entire goal of pathfinding to set proper routes between agents and targets? Why do we need this `nearby_walkable` function...?

**Comment by Libbum** [comment]
In addition to George's latest question, I'd like to point out that

`walk!(agent, rand, model)`

is not really a 'random walk' in the way that term is usually understood: it's a movement in a random direction that ultimately pulls from a uniform distribution. 

Sorry, I've not had the time to look at the code yet, so I think I'm missing the thread here too.

**Comment by AayushSabharwal** [comment]
`nearby_walkable` is a utility function. While making the rabbit/fox/hawk example, I ended up doing the same thing repeatedly and made it a function. It seemed like a useful function to have for grid space pathfinding.

This week, @Libbum and I discussed that it might be useful for rabbit-fox-hawk to also have a version in continuous space (similar to how there are multiple versions of predator-prey and daisyworld), as an example of why one might need it (having the concept of speed). I was going through the code for that example and checking if there's some functionality related to pathfinding in grid space that didn't have a direct alternative in continuous space, and this jumped out. Currently the method used for random walks in the example isn't a true random walk either. For that, I pick a random position in the vision radius and move there. Since the radius is large enough I think it works as a decent approximation. Replicating it for continuous space isn't easily possible with the current API

**Comment by Libbum** [comment]
OK, thanks for clearing that up. I'll come back to you tomorrow morning on this - will do a proper review then.

**Comment by AayushSabharwal** [comment]
I've compiled a list of all functions I think can be slightly modified to work with pathfinders:
- [ ] `random_position(model, pathfinder)` - Considers walkability of the returned position ⭐ 
- [ ] `walk!(agent, rand, model, pathfinder)` - Considers walkability of the random position
- [ ] `positions(model, pathfinder [, by])` - For discrete space, only returns walkable positions
- [ ] `fill_space!([A, ] model, args...; kwargs...)` and `fill_space!([A ,] model, f; kwargs...)` - Only fills walkable positions
- [ ] `nearby_positions(positions, model, r = 1)` - Returns only walkable positions ⭐ 
- [ ] `add_agent!([A, ] model, pathfinder, args...; kwargs...)` - The agent is added to a random walkable position ⭐ 
- [ ] The empty positions related methods - Consider walkability of the returned positions

I've marked with ⭐ the ones I think are more important

**Comment by Datseris** [comment]
I vote that fundamentally central functions like `add_agent!` and `random_position` should **not** have overloads w.r.t. pathfinder, which I still consider a minor functionality. Sure, it took huge effort, and many lines of code to write, but if one consider all possible ABMs, pathfinding is minor while `add_agent!` is used in every single one.

Just notice: multiple dispatch is awesome, but everything should be taken **in moderation**. The more methods you add to `add_agents!`, the harder it becomes to learn. So you need to weight the generality of the new added method, versus the increase it gives in the methods of `add_agent!`.

Why not have `random_walkable` and `nearby_walkable` as different names instead? The function `add_agent!` can then just use the result of `random_walkable`.

**Comment by Datseris** [comment]
Also, I'm completely against having "walk! with pathfinder" and I'm still unconvinced of its use. The whole purpose is setting routes via the standard approach. Why is walk here I still don't see. Just pick a random `random_walkable` and set a route to that. Perhaps you can add some kind of limiter to `random_walkable` to not give a grid cell too far away or something like that.

**Comment by AayushSabharwal** [comment]
I see your point now. If all we're doing is respecting walkability, we can provide those two functions and leave it to the user to use them. The other alternative functions I proposed would boil down to simple wrappers over `random_walkable`/`nearby_walkable`, which isn't worth the added complexity, both in terms of the number of methods scattered around and documentation. I'll add these two new functions.

I think both of them can be extended to continuous space fairly easily. All it would require is generating a random grid cell in the walkable matrix and taking a random poisition inside that cell. `nearby_walkable` would additionally do a length-check on the vector, since grid cells at the edge would also have points outside the radius.

**Comment by AayushSabharwal** [comment]
Although, `nearby_walkable` in continuous space wouldn't benefit from grid space's nearest neighbor optimisations, specifically the caching of hoods.

**Comment by AayushSabharwal** [comment]
I think this closes the list of features that continuous pathfinding needed. I'd say it's ready for review now.

**Comment by Datseris** [comment]
@AayushSabharwal this probably comes as a late notice... but can you prepare a small high level summary for the GSOC meeting tomorrow (I'll also join)?

File: src/submodules/pathfinding/astar_continuous.jl
Comment (Datseris) [inline src/submodules/pathfinding/astar_continuous.jl:3]: instead of find_path, type `find_path` (i.e. anything that is _code_ must be in backticks)

**Comment by AayushSabharwal** [comment]
I just noticed a bug: In `set_best_target!` for a grid space, there is no differentiation between the current position of the agent and unreachable positions. i.e. if I pass `[agent.pos, some_unreachable_position]` to it, it will return `some_unreachable_position` since the length of the returned path to both is 0
This doesn't happen with continuous space, since `find_continuous_path` returns `nothing` if no path exists. I'll modify `find_path` to do this as well, and `set_best_target!` such that it can handle such a scenario

**Comment by AayushSabharwal** [comment]
The second implementation of `random_walkable` for continuous space ended up having a lot of bugs and nuances associated with it. They should be fixed now

**Comment by AayushSabharwal** [comment]
I'm not fully sure how to write tests for an RNG based function, since their output is inherently unpredictable. For now, I've taken 10 samples for each test. However, this still likely won't catch every scenario. I didn't seed the model RNG since it's possible some other seed could cause an issue that the one I use won't.

**Comment by Libbum** [comment]
Yeah, they are fundamentally a problem. Just do the best you can. A sense check on a seed is fine, make sure that some property that comes out (even just the size of a vector for example) is the same.

To be completely happy with most of Agents.jl, we'd need to do some fuzz testing, so don't fret too much on having poor unit tests in this area.

**Comment by Datseris** [comment]
should we merge this into the pathfinding branch and give a final review?

File: CHANGELOG.md
Comment (Libbum) [inline CHANGELOG.md:5]: New functions `get_spatial_property` and `get_spatial_index`, which provide better handling of discrete representations of spatial fields present in `ContinuousSpace`.

File: docs/src/comparison.md
Comment (Libbum) [inline docs/src/comparison.md:15]: Why is this file here? This should already be in master?

File: src/Agents.jl
Comment (Libbum) [inline src/Agents.jl:3]: Same deal here. Looks like you're out of sync a little.

File: CHANGELOG.md
Comment (Datseris) [reply CHANGELOG.md:5]: that's already from my PR

**Review by Libbum — APPROVED** [review]
Fine to merge into the rework branch here, there are some sync issues that are probably and artifact of branching from different locations, so that should be fine.

Don't forget to remove the `- pathfinding_rework` addition to `.github/workflows/ci.yaml`

**Comment by AayushSabharwal** [comment]
Is there any reason to squash here, since we're not merging to master and maintaining history could be useful?

**Comment by Datseris** [comment]
I have revisited this request on merging commits and did some research on good practices on collaborative code projects. The conclusion is: **always squash**. @AayushSabharwal I'm not sure where you heard that "you don't take credit" if PRs are squashed instead of creating a merge commit, but it is simply not true. Furthermore I believe that acknowledging contributions should be done in a dedicated place in the repo/docs, and not reliant on GitHub's statistics. I'd recommend also watching Bogumil's JuliaCon talk https://pretalx.com/juliacon2021/talk/RJE93F/

**Pull request merged**
